tests/integration/security/ca_custom_root/trust_domain_alias_client_test.go

```diff
@@ -0,0 +1,182 @@
+			var client, serverNakedFoo, serverNakedBar echo.Instance
+			echoboot.NewBuilder(ctx).
+				With(&client, echo.Config{
+					Namespace: testNS,
+					Service:   "client",
+				}).
+				With(&serverNakedFoo, echo.Config{
+					Namespace: testNS,
+					Service:   "server-naked-foo",
+					Subsets: []echo.SubsetConfig{
+						{
+							Annotations: echo.NewAnnotations().SetBool(echo.SidecarInject, false),
+						},
+					},
+					ServiceAccount: true,
+					Ports: []echo.Port{
+						{
+							Name:         HTTPS,
+							Protocol:     protocol.HTTPS,
+							ServicePort:  443,
+							InstancePort: 8443,
+							TLS:          true,
+						},
+					},
+					TLSSettings: &common.TLSSettings{
+						RootCert:   loadCert(t, "root-cert.pem"),
+						ClientCert: loadCert(t, "tmp/workload-foo-cert.pem"),
+						Key:        loadCert(t, "tmp/workload-foo-key.pem"),
+					},
+				}).
+				With(&serverNakedBar, echo.Config{
+					Namespace: testNS,
+					Service:   "server-naked-bar",
+					Subsets: []echo.SubsetConfig{
+						{
+							Annotations: echo.NewAnnotations().SetBool(echo.SidecarInject, false),
+						},
+					},
+					ServiceAccount: true,
+					Ports: []echo.Port{
+						{
+							Name:         HTTPS,
+							Protocol:     protocol.HTTPS,
+							ServicePort:  443,
+							InstancePort: 8443,
+							TLS:          true,
+						},
+					},
+					TLSSettings: &common.TLSSettings{
+						RootCert:   loadCert(t, "root-cert.pem"),
+						ClientCert: loadCert(t, "tmp/workload-bar-cert.pem"),
+						Key:        loadCert(t, "tmp/workload-bar-key.pem"),
+					},
+				}).
+				BuildOrFail(t)
+
+			ctx.Config().ApplyYAMLOrFail(ctx, testNS.Name(), fmt.Sprintf(POLICY))
+
+			verify := func(t *testing.T, src echo.Instance, dest echo.Instance, s scheme.Instance, allow bool) {
+				t.Helper()
+				want := "allow"
+				if !allow {
+					want = "deny"
+				}
+				name := fmt.Sprintf("server:%s[%s]", dest.Config().Service, want)
+				t.Run(name, func(t *testing.T) {
+					t.Helper()
+					opt := echo.CallOptions{
+						Target:   dest,
+						PortName: HTTPS,
+						Address:  dest.Config().Service,
+						Scheme:   s,
+					}
+					checker := connection.Checker{
+						From:          src,
+						Options:       opt,
+						ExpectSuccess: allow,
+					}
+					checker.CheckOrFail(ctx)
+				})
+			}
+
+			verify(t, client, serverNakedFoo, scheme.HTTP, true)
+			verify(t, client, serverNakedBar, scheme.HTTP, false)
+		})
+}
+
+func loadCert(t test.Failer, name string) string {
+	data, err := cert.ReadSampleCertFromFile(name)
+	if err != nil {
+		t.Fatal(err)
+	}
+	return string(data)
+}
+
+func generateCerts(ctx framework.TestContext, ns string, td string, sa string) {
+	workDir := path.Join(env.IstioSrc, "samples/certs")
+	script := path.Join(workDir, "generate-workload.sh")
+	command := exec.Cmd{
+		Path:   script,
+		Args:   []string{script, td, ns, sa, workDir, "tmp"},
```

Comment:
You can make cleanup function as a returned value. and the caller would just doing 

```
cleanup := generateCerts()
defer cleanup()
```

This makes logic tigher. e.g. you don't have to remember `tmp` folder is copied in two places.

Reply:
updated